### PR TITLE
feat: add serving size validation using units taxonomy

### LINKED ISSUE #807: Serving Size Validation
Serving Size Validation - Perform real-time validation on the serving size field to ensure it contains a number and a valid unit (e.g., `100 g`, `250 ml`).

Image
(not the right photo)

### PR DESCRIPTION
**Pull request by bobcorn**
## Description

This PR adds serving size validation using the `units` taxonomy as the source of truth.

The validation is non-blocking and provides feedback according to the following logic:
<table>
  <tr>
    <th align="center">Validation</th>
    <th align="center">Cases</th>
  </tr>
  <tr>
    <td valign="middle" align="center"><strong>No issue</strong></td>
    <td valign="top">
      <ul>
        <li>Input is empty</li>
        <li>Input contains a number and a recognized unit</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center"><strong>Error</strong></td>
    <td valign="top">
      <ul>
        <li>Input is not empty but missing a number</li>
        <li>Input contains a number but missing a unit</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center"><strong>Warning</strong></td>
    <td valign="top">
      <ul>
        <li>Input contains a number and text that could be a unit, but the unit is not recognized</li>
      </ul>
    </td>
  </tr>
</table>

As a conservative approach, inputs are considered valid whether the unit appears before or after the number, to avoid confusing users when they provide imperfect but still useful inputs like `ml 250`.

## Changes
<table>
  <tr>
    <th align="center">File</th>
    <th align="center">Changes</th>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-7020a6bab670b4f20815023475f0a0436f9092968813d5e62e81d4be919c3d62"><code>types.ts</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Added <code>Unit</code> taxonomy type</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-2ed210d24f6b0e53042fff18a33fac0d46b6634718cdb54982c391712c8e7ecc"><code>+page.ts</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Added <code>units</code> taxonomy to page load data</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-4fa0132850ff7bb5564a3d2d9c412093d3d20e4e49debd9f2f9c09909fa7ebb2"><code>AddProductForm.svelte</code></a><br/>
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-0ef0b341f6c7940ea4c0f823dc1661eacb7d9c1817812512fce53b1576901dbd"><code>EditProductForm.svelte</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Passed units data to the nutrition step</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-d6e1b392c45cf62bae7e52eae7199d171234593fb94832141acf5ade3c74cb80"><code>+page.svelte</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Unpacked <code>units</code> taxonomy as a list of strings</li>
        <li>Passed units to the edit form flow</li>
        <li>Added sizing class to warning icon in dev alert banner as it was showing super big (this was not in the scope of the PR, but was a very small fix so I thought reasonable to include it)</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-a67da16d0bd68853f5e4a4d2db00b5dc0ea9c2830ea9ea948f830fea8ecef2a1"><code>nutrition.ts</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Added serving size validation logic</li>
        <li>Reused same issue model already used by nutrition step</li>
        <li>Renamed some shared types and functions so they are now consistent with both serving size validation and nutrition analysis</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-c7036034ed803ec596aecf37b3e201e12bb7be0ff040f4703051a35a60588757"><code>NutritionStep.svelte</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Added serving size validation in nutrition step UI</li>
        <li>Reused shared tooltip and alert rendering for both serving size and existing nutrition issues</li>
        <li>Kept serving size feedback independent from <code>no_nutrition_data</code></li>
      </ul>
    </td>
  </tr>
  <tr>
    <td valign="middle" align="center">
      <a href="https://github.com/openfoodfacts/openfoodfacts-explorer/pull/1209/changes#diff-38b445b65396e04aa4d1f617391a87751acc427fe493df0996a56b41a3cf77b7"><code>nutrition.test.ts</code></a>
    </td>
    <td valign="top">
      <ul>
        <li>Added unit tests for serving size validation</li>
      </ul>
    </td>
  </tr>
</table>

## Related issue
Fixes #807

## Notes
- As discussed in the related issue, I initially considered lazy-loading the units taxonomy on field focus, but ultimately I realized it was more consistent to align it with the existing taxonomy loading flow used by the edit page.

## Screenshots
<table>
  <tr>
    <td colspan="6" align="center">
      <strong>Serving size validation</strong>
    </td>
  </tr>
  <tr>
    <td colspan="2" align="center" valign="top">
      <strong>No input</strong><br/>
      <a href="https://github.com/user-attachments/assets/2ee309f2-0276-4580-9f6c-db221599f886" target="_blank">
        
      </a>
    </td>
    <td colspan="2" align="center" valign="top">
      <strong>Valid input</strong><br/>
      <a href="https://github.com/user-attachments/assets/646dcaba-83ff-4c66-aca6-53b0b743babb" target="_blank">
        
      </a>
    </td>
    <td colspan="2" align="center" valign="top">
      <strong>Missing number</strong><br/>
      <a href="https://github.com/user-attachments/assets/7233d08e-c85d-4f59-b3dd-03be13d6f7a7" target="_blank">
        
      </a>
    </td>
  </tr>
  <tr>
    <td colspan="2" align="center" valign="top">
      <strong>Missing unit</strong><br/>
      <a href="https://github.com/user-attachments/assets/18287ee9-d98e-471e-bfdb-ba1776f9beed" target="_blank">
        
      </a>
    </td>
    <td colspan="2" align="center" valign="top">
      <strong>Unrecognized unit</strong><br/>
      <a href="https://github.com/user-attachments/assets/0c2ce146-00ee-4189-a627-986e2078ee74" target="_blank">
        
      </a>
    </td>
    <td colspan="2" align="center" valign="top">
      <strong>No nutrition data</strong><br/>
      <a href="https://github.com/user-attachments/assets/02387e8a-7442-4ff0-9666-b0e459cdd081" target="_blank">
        
      </a>
    </td>
  </tr>

  <tr>
    <td colspan="6" align="center">
      <strong>Warning icon bug</strong>
    </td>
  </tr>
  <tr>
    <td colspan="3" align="center" valign="top">
      <strong>Before</strong><br/>
      <a href="https://github.com/user-attachments/assets/f45ce574-469a-470c-988e-df243434ff17" target="_blank">
        
      </a>
    </td>
    <td colspan="3" align="center" valign="top">
      <strong>After</strong><br/>
      <a href="https://github.com/user-attachments/assets/d05a3178-1e74-4434-909e-96f9b75abe2d" target="_blank">
        
      </a>
    </td>
  </tr>
</table>

## Summary by CodeRabbit

* **New Features**
  * Product forms now receive unit data so serving-size fields recognize unit names and symbols.
  * Pages now load and pass unit lists to forms.

* **Improvements**
  * Serving-size validation added with clearer severity-aware alerts/tooltips and refined input styling.
  * Nutrition issue reporting consolidated and more consistent.

* **Localization**
  * New localized strings for serving-size examples, placeholders, and validation messages (en/it).

* **Tests**
  * Added tests covering serving-size parsing and unit-recognition edge cases.

### CODE REVIEW

File: src/routes/products/[barcode]/edit/+page.svelte [hunk 1]
```diff
@@ -139,12 +140,35 @@
 			.filter((t): t is string => t !== undefined);
 	}
 
+	function getUnits(taxo: Taxonomy<Unit>) {
+		const units: string[] = [];
+		const seenUnits: Record<string, true> = {};
```
[the remark this answers is not in the document]
Comment (bobcorn) [reply src/routes/products/[barcode]/edit/+page.svelte]: You're right, fixed!

File: src/lib/ui/edit-product-steps/nutrition.ts [hunk 2]
```diff
@@ -1,13 +1,119 @@
...
+	const tokens = normalizedServingSize.split(NUMBER_PATTERN);
+	let containsUnknownUnit = false;
+
+	for (const token of tokens) {
+		if (token.trim() === '') {
+			continue;
+		}
+
+		if (containsKnownUnit(token, units)) {
+			return 'valid';
```
Comment (VaiTon) [inline src/lib/ui/edit-product-steps/nutrition.ts:71]: This returns 'valid' eagerly on the first token that contains a known unit, which means that something like `34 g asdasd` would be valid.

File: src/lib/ui/edit-product-steps/nutrition.ts [hunk 3]
```diff
@@ -1,13 +1,119 @@
...
+			};
+		case 'unknown-unit':
+			return {
+				severity: 'warning',
+				field: 'serving_size',
+				title: 'Serving size unit may be incorrect',
+				desc: `Serving size should include a common unit. ${servingSizeExamples}`
+			};
+	}
+}
```
Comment (VaiTon) [inline src/lib/ui/edit-product-steps/nutrition.ts]: This should be all i18n-able

File: src/lib/ui/edit-product-steps/NutritionStep.svelte [hunk 4]
```diff
@@ -81,29 +82,58 @@
...
 		if (a.severity === b.severity) return 0;
 		if (a.severity === 'error') return -1;
 		return 1;
 	};
 
+	function inputClassForSeverity(severity: 'warning' | 'error' | undefined): string {
+		if (severity === 'error') return 'input-error';
+		if (severity === 'warning') return 'input-warning';
+		return '';
+	}
```
Comment (VaiTon) [inline src/lib/ui/edit-product-steps/NutritionStep.svelte]: Make this a const object.

File: src/lib/ui/edit-product-steps/NutritionStep.svelte [hunk 5]
```diff
@@ -81,29 +82,58 @@
...
+
+	function highestSeverity(
+		issues: Array<{ severity: 'warning' | 'error' }>
+	): 'warning' | 'error' | undefined {
+		if (issues.some((r) => r.severity === 'error')) return 'error';
+		if (issues.some((r) => r.severity === 'warning')) return 'warning';
+		return undefined;
+	}
```
Comment (VaiTon) [inline src/lib/ui/edit-product-steps/NutritionStep.svelte]:
```
const order = [error, warning]
for (sev of order) {
   if (issues.some(...)) return sev;
}
```

**Review by VaiTon — COMMENTED** [review]
Sorry for the delay!

**Comment by bobcorn** [comment]
> Sorry for the delay!

No worries at all, thanks for taking the time! I’ll go through all the comments

File: src/lib/ui/edit-product-steps/NutritionStep.svelte (hunk 4 above)
Comment (bobcorn) [reply src/lib/ui/edit-product-steps/NutritionStep.svelte]: Done!

File: src/lib/ui/edit-product-steps/NutritionStep.svelte (hunk 5 above)
Comment (bobcorn) [reply src/lib/ui/edit-product-steps/NutritionStep.svelte]: Updated this to use a precedence array

File: src/lib/ui/edit-product-steps/nutrition.ts (hunk 3 above)
Comment (bobcorn) [reply src/lib/ui/edit-product-steps/nutrition.ts]: Moved the serving size messages into i18n

File: src/lib/ui/edit-product-steps/nutrition.ts (hunk 2 above)
Comment (bobcorn) [reply src/lib/ui/edit-product-steps/nutrition.ts:71]: I treated this as valid because the original placeholder for the input field was `e.g. 100g, 1 serving (30g)`, so invalidating `34 g asdasd` would likely invalidate that format too. Should we keep this permissive, or tighten it and change the placeholder?

**Comment by bobcorn** [comment]
@VaiTon pushed an update to address the comments. I only left the open point around `34 g asdasd` validation, since it probably needs some clarification.

File: src/lib/ui/edit-product-steps/nutrition.ts (hunk 2 above)
Comment (VaiTon) [reply src/lib/ui/edit-product-steps/nutrition.ts:71]: You got a point. Let's keep it like it is right now.

**Review by VaiTon — APPROVED** [review]
LGTM
